File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
#include <stack>

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int mx = 0;
        stack<int> stack;
        for(int i=0; i<prices.size(); i++){
            if(stack.empty()){
                stack.push(prices[i]);
            }else{
                if(stack.top() > prices[i]){
                    stack.push(prices[i]);
                }else{
                    if(mx < prices[i] - stack.top()){
                        mx = prices[i] - stack.top();
                    }
                }
            }
        }
        return mx;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
import java.util.Stack;

class Solution {
    public int maxProfit(int[] prices) {
        Stack<Integer> st = new Stack<>();
        
        int mx = 0;
        for(int i=0; i<prices.length; i++){
            if(st.isEmpty()){
                st.add(prices[i]);
            }else{
                if(st.peek() > prices[i]){
                    st.add(prices[i]);
                }else{
                    if(mx < prices[i] - st.peek()){
                        mx = prices[i] - st.peek();
                    }            
                }
            }
        }
        return mx;
    }
}